fission-web-server/library/Fission/Web/Server/RecoveryChallenge.hs

```diff
@@ -0,0 +1,20 @@
+module Fission.Web.Server.RecoveryChallenge
+  ( module Fission.Web.Server.RecoveryChallenge.Creator.Class
+  , module Fission.Web.Server.RecoveryChallenge.Retriever.Class
+  , recoveryLink
+  ) where
+
+import           Servant                                              hiding
+                                                                      (route)
+
+import           Fission.Prelude
+
+import           Fission.Challenge.Types
+
+import           Fission.Web.Server.RecoveryChallenge.Creator.Class
+import           Fission.Web.Server.RecoveryChallenge.Retriever.Class
+
+
+recoveryLink :: Text -> Challenge -> Text
+recoveryLink recoveryAppUrl challenge =
+  recoveryAppUrl <> "?challenge=" <> toUrlPiece challenge
```

Comment:
Could we generate this with `mkLink`?

Reply:
I've [tried that before](https://github.com/fission-suite/fission/pull/529/commits/3d9351bb8053d7a4110c244c33495fabaae525ed#diff-f1e24b239e27892463f95beea5b05887f5d6be9fe542ba2e4f48ade5f3e47f16R18-R25), but it felt kind of clunky, because it's not part of the actual server API, but an 'apps API'. Also, it separates the baseUrl from the route, and the route itself is hardcoded, so instead of `recoveryAppUrl` being configurable as e.g. `https://dashboard.fission.codes/recover/`, the `recover/` part would have to be hardcoded in a servant route type.
Aaand I didn't know how to generate a link with a trailing slash (which I need unfortunately) using servant's `safeLink`. (Maybe `recover :> "" :> QueryParam ...` would've worked?)

If there's a more structural way to create http urls somehow, I'd love to use them (something like [`elm/url`](https://package.elm-lang.org/packages/elm/url/latest/Url)?)

Reply:
Oh, I see, this is the link to the dashboard app. Yeah makes sense, unless we start having a bunch of these.

> with a trailing slash (which I need unfortunately) 

Why? Is this something that Elm expects? Or is it an implicit `index.html`?

> If there's a more structural way to create http urls somehow, I'd love to use them (something like elm/url?)

Yeah, there's definitely options for these:

1. Create a `data TrailingSlash`, and give it the required Servant instances
2. Use [`URI`](https://www.stackage.org/haddock/lts-18.0/network-uri-2.6.4.1/Network-URI.html#v:uri)
  * Can then plug into `linkURI :: Link -> URI`, so something like `withTrailing = show . addSlash . linkURI . mkLink`

Reply:
> Why? Is this something that Elm expects? Or is it an implicit index.html?

Yeah, there's an index.html file at https://matheus23-dashboard.fission.pizza/recover/index.html

---

Thanks for the pointers to URI/linkURI!